presto-main/src/main/java/com/facebook/presto/operator/window/ReflectionWindowFunctionSupplier.java

```diff
@@ -31,7 +33,15 @@
 public class ReflectionWindowFunctionSupplier<T extends WindowFunction>
         extends AbstractWindowFunctionSupplier
 {
+    private enum ConstructorType
```

Comment:
I think `ignoreNulls` should be a variable in `ValueWindowFunction`. You can have a separate `setIgnoreNulls` in `ValueWindowFunction`, that way you don't need to mess with the constructors. `ignoreNulls` is not really part of the function inputs anyways. If you keep a separate `currentNonNullPosition` in `ValueWindowFunction`, I believe the null handling can be done in `ValueWindowFunction.processRow`.

Reply:
I have made the suggested change with `ignoreNulls` - looking into other suggested change.

Reply:
I'm looking into the other suggested change to move null handling to `ValueWindowFunction.processRow`.

Reply:
You probably still need special handling on each function, but how many non-null values there are since `frameStart` can be tracked. The current implementation is really inefficient as it tries to iterate from the beginning of the frame on every position to count whether it got enough non-null values. This is O(n^2) while you could have tracked this in O(n).